### PR TITLE
Generalize CSP headers

### PR DESCRIPTION
**Pull request by nmattia**
This updates the way we create the CSP headers (and in particular, the `script-src` clause).

Before this, and due to a (now resolved) bug in Firefox, we could not specify `integrity=` attributes for whitelisting scripts in HTML. Instead we implemented a work around where we injected a new inline script (with known hash) which in turn loaded the actual JS script.

This was unconventional, somewhat unsafe (because it didn't actually include the script's hash) and brittle (we hard coded some script names, like `index.js`, `index2.js`, etc).

This updates the frontend bundle to include `integrity=` attributes in HTML script tags, and updates CSP header creation to extract those integrity hashes and whitelist them in the generated asset list.

The CSP header creation is also changed to keep a reference to the security headers certified alongside the assets. This ensures the security headers do not need to be recomputed with every asset, and clarifies that assets were certified with specific headers.

The CSP header error handling in the test suite is updated for better error reporting.

Firefox references:
* https://bugzilla.mozilla.org/show_bug.cgi?id=1409200
* https://www.mozilla.org/en-US/firefox/116.0a1/releasenotes/

<hr/><details><summary>🟡 Some screens were changed</summary></details>

### CODE REVIEW

File: src/vite-plugins/src/index.ts [hunk 1]
```diff
@@ -139,3 +141,53 @@ export const replicaForwardPlugin = ({
     });
   },
 });
+
+/** Update the HTML files to include integrity hashes for script.
+ * i.e.: `<script src="foo.js">` becomes `<script integrity="<hash(./foo.js)>" src="foo.js">`.
+ */
+export const integrityPlugin: Plugin = {
```
Comment (frederikrothenberger) [inline src/vite-plugins/src/index.ts:148]: It's quite surprising to me that there is no ready-made plugin for that. Or do we have additional requirements to build our own?

Comment (nmattia) [reply src/vite-plugins/src/index.ts:148]: There is some, but they pull in an HTML parser and did not work for me

File: src/internet_identity/src/http.rs [hunk 2]
```diff
@@ -197,12 +198,23 @@ pub fn content_security_policy_header() -> String {
     csp
 }
 
-fn certified_asset(asset_name: &str, certificate_version: Option<u16>) -> Option<CertifiedAsset> {
+/// Read an asset from memory, returning the associated HTTP code, content and full list of
+/// headers that were certified with the asset.
+fn get_asset(
```
[the remark this answers is not in the document]
Comment (nmattia) [reply src/internet_identity/src/http.rs:203]: I thought about that at first and then found it confusing that sometimes the `CertifiedAsset` includes the shared headers, sometimes not. WDYT?

[the remark this answers is not in the document]
Comment (frederikrothenberger) [reply src/internet_identity/src/http.rs:203]: Simplest would be to just remove the concept of `shared_headers` altogether and just duplicate them among all assets.
I don't think we are resource constrained enough to justify the complexity of deduplication. But that should probably go into a different PR.

File: src/vite-plugins/src/index.ts (hunk 1 above)
Comment (frederikrothenberger) [reply src/vite-plugins/src/index.ts:148]: Thanks for the clarification.

File: src/internet_identity/src/http.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (nmattia) [reply src/internet_identity/src/http.rs:203]: > I don't think we are resource constrained enough to justify the complexity of deduplication. 

I agree

**Pull request merged**